### PR TITLE
Simplex tree dimension simplex range

### LINKED ISSUE #183: [Simplex_tree] dimension_simplex_range
Sometimes we want to iterate on all simplices of dimension exactly k. Currently, that means iterating on the k-skeleton, and checking for each simplex if its dimension is k. It would be convenient (and a bit cheaper) to have an iterator that directly gives us the right simplices.

### PR DESCRIPTION
**Pull request by VincentRouvreau**
* Fix #183 
* Use it in Alpha complex and some unitary tests

TBD: python interfaces

### Bench results
```txt
$ ./simplex_tree_dimension_simplex_range_benchmark
... Simplex_tree construction
... 96951 vertices and 1147249 simplices. Dimension is 4
... benchmark for dimension 0
Benchmark the skeleton range but with the exact dim test: 0.001s
 - count = 96951
Benchmark the dimension range: 0s
 - count = 96951
... benchmark for dimension 1
Benchmark the skeleton range but with the exact dim test: 0.023s
 - count = 500280
Benchmark the dimension range: 0.018s
 - count = 500280
... benchmark for dimension 2
Benchmark the skeleton range but with the exact dim test: 0.037s
 - count = 375156
Benchmark the dimension range: 0.034s
 - count = 375156
... benchmark for dimension 3
Benchmark the skeleton range but with the exact dim test: 0.042s
 - count = 149923
Benchmark the dimension range: 0.037s
 - count = 149923
... benchmark for dimension 4
Benchmark the skeleton range but with the exact dim test: 0.042s
 - count = 24939
Benchmark the dimension range: 0.038s
 - count = 24939
... benchmark for dimension 5
Benchmark the skeleton range but with the exact dim test: 0.043s
 - count = 0
Benchmark the dimension range: 0s
 - count = 0
... benchmark for dimension 6
Benchmark the skeleton range but with the exact dim test: 0.042s
 - count = 0
Benchmark the dimension range: 0s
 - count = 0
```

### CODE REVIEW

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h [hunk 1]
```diff
@@ -455,6 +455,102 @@ class Simplex_tree_skeleton_simplex_iterator : public boost::iterator_facade<
   int curr_dim_;
 };
 
+/** \brief Iterator over the simplices of the simplicial complex that match the dimension specified by the parameter.
+ *
+ * Forward iterator, value_type is SimplexTree::Simplex_handle.*/
+template<class SimplexTree>
+class Simplex_tree_dimension_simplex_iterator : public boost::iterator_facade<
```
Comment (hschreiber) [inline src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h:462]: I feel like `Simplex_tree_complex_simplex_iterator`, `Simplex_tree_skeleton_simplex_iterator` and `Simplex_tree_dimension_simplex_iterator` are more or less the same class with just different stopping conditions. I am wondering if they would be a clean way to factorize them? This would also highlight more the difference between them (codewise).

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h [hunk 2]
```diff
@@ -455,6 +455,102 @@ class Simplex_tree_skeleton_simplex_iterator : public boost::iterator_facade<
...
+      return false;
+    }
+    return (&(sh_->second) == &(other.sh_->second));
+  }
+
+  Simplex_handle const& dereference() const {
+    return sh_;
+  }
+
+  inline void until_leaf_or_dim() {
```
Comment (hschreiber) [inline src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h]: `inline` should be implicit here, it does not need to be specified.

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 3]
```diff
@@ -351,6 +357,18 @@ class Simplex_tree {
                                   Skeleton_simplex_iterator());
   }
 
+  /** \brief Returns a range over the simplices the simplicial complex that match the dimension specified by the
```
Comment (hschreiber) [inline src/Simplex_tree/include/gudhi/Simplex_tree.h]:
```suggestion
  /** \brief Returns a range over the simplices of the simplicial complex that match the dimension specified by the
```

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h [hunk 4]
```diff
@@ -455,6 +455,102 @@ class Simplex_tree_skeleton_simplex_iterator : public boost::iterator_facade<
   int curr_dim_;
 };
 
+/** \brief Iterator over the simplices of the simplicial complex that match the dimension specified by the parameter.
```
Comment (hschreiber) [inline src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h:458]:
```suggestion
/** \brief Iterator over the simplices of the simplicial complex that matches the dimension specified by the parameter.
```

Comment (mglisse) [inline src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h:458]: Why? "the simplices [...] match" looks correct.

File: src/Alpha_complex/concept/SimplicialComplexForAlpha.h [hunk 5]
```diff
@@ -66,10 +66,9 @@ struct SimplicialComplexForAlpha {
   /** \brief Iterator over the simplices of the skeleton of the complex, for a given dimension.
    *
    * 'value_type' must be 'Simplex_handle'. */
-  typedef unspecified Skeleton_simplex_range;
-  /** \brief Returns a range over the simplices of the skeleton of the simplicial complex, for a given
-   * dimension. */
-  Skeleton_simplex_range skeleton_simplex_range;
+  typedef unspecified Dimension_simplex_range;
+  /** \brief Returns a range over the the simplices of the simplicial complex for a given dimension. */
+  Dimension_simplex_range dimension_simplex_range(int dimension);
```
Comment (mglisse) [inline src/Alpha_complex/concept/SimplicialComplexForAlpha.h:71]: This is an API break, so it needs to be mentioned in the release notes.

File: src/Simplex_tree/benchmark/random_simplices.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+#include <iostream>
+#include <random>
+#include <numeric>  // for std::iota
+#include <algorithm>  // for std::shuffle
+#include <vector>
+
+std::random_device rd;
+
+template <class Vertex_handle>
+std::vector<Vertex_handle> rand_int_range(int subset_min_size,
```
Comment (mglisse) [inline src/Simplex_tree/benchmark/random_simplices.h]: Add a comment above explaining what this computes? (or give the function a name that makes it obvious, `rand_int_range` doesn't)

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h [hunk 7]
```diff
@@ -455,6 +455,102 @@ class Simplex_tree_skeleton_simplex_iterator : public boost::iterator_facade<
...
+        dim_(0),
+        curr_dim_(0) {
+  }
+
+  Simplex_tree_dimension_simplex_iterator(SimplexTree const* st, int dim)
+      : sib_(nullptr),
+        st_(st),
+        dim_(dim),
+        curr_dim_(0) {
+    if (st == nullptr || st->root() == nullptr || st->root()->members().empty() || st->dimension() < dim) {
```
Comment (mglisse) [inline src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h]:
```suggestion
    if (st == nullptr || st->root() == nullptr || st->root()->members().empty() || st->upper_bound_dimension() < dim) {
```
?
If we ask for the 0-skeleton, it feels wrong to iterate through the whole complex to compute its dimension before we start iterating on the vertices.
Ah, maybe the code that follows assumes that the range is not empty and misbehaves if it is, but then maybe that can be improved?

Should we also test `dim<0` here?

Is there a unit test for the empty complex?

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h (hunk 4 above)
Comment (hschreiber) [reply src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h:458]: I read this sentence three times to be sure and read a singular every time... You are right, I will delete the comment.

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h (hunk 2 above)
Comment (VincentRouvreau) [reply src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h]: I removed it on 6fee5ca

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h (hunk 7 above)
Comment (VincentRouvreau) [reply src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h]: test 'dim<0' and add unitary test on an empty complex on 4a4e896

File: src/Alpha_complex/concept/SimplicialComplexForAlpha.h (hunk 5 above)
Comment (VincentRouvreau) [reply src/Alpha_complex/concept/SimplicialComplexForAlpha.h:71]: I added the API break message on [ef33f4d](https://github.com/GUDHI/gudhi-devel/pull/1229/commits/ef33f4d0c1e45b298fb0475a3237747554a63261)

File: src/Simplex_tree/benchmark/random_simplices.h (hunk 6 above)
Comment (VincentRouvreau) [reply src/Simplex_tree/benchmark/random_simplices.h]: I renamed `rand_int_range` as `random_simplex` and added some internal doc on [958156b](https://github.com/GUDHI/gudhi-devel/pull/1229/commits/958156bae11386ec2b9a8f2104115cd18d221711)

File: src/Simplex_tree/benchmark/random_simplices.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <iostream>
+#include <random>
+#include <numeric>  // for std::iota
+#include <algorithm>  // for std::shuffle
+#include <vector>
+
+std::random_device rd;
+
+/** \brief Returns a random simplex.
```
Comment (mglisse) [inline src/Simplex_tree/benchmark/random_simplices.h:19]: I was afraid doxygen might try to show it, but I don't see it in the generated doc, that's good.

File: src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h (hunk 1 above)
Comment (mglisse) [reply src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h:462]: I agree that it would be nice to factorize them. I think we can introduce the new interface with duplicated code now and delay any refactoring, but if you disagree we can block the PR until it is done.

(it is a shame that C++ coroutines aren't better, it would be so much more readable to implement this with for_each_simplex + yield)

**Review by mglisse — APPROVED** [review]
